### DISCUSSION TITLE
How to verify if an xpub signed a psbt?

### DISCUSSION THREAD

**Discussion by sebastianmontero** [rust-bitcoin/rust-bitcoin#1070] [Q&A] [answered]
Given an xpub and a psbt is there a way to check if the xpub is one of the signers of the psbt?

**Comment by DanGould** [comment, marked as answer]
Given an xpub alone, no, not easily.

Given an xpub and a derivation path, yes: Derive public keys according to the derivation path and check them against the public keys in psbt output scripts. Then verify the signature. This is how psbt signing works in hardware wallets available today.

Note that derivation is tricky because keys can be generated a near-infinite number of ways. So one must recall exactly how the keys were derived in order to find them from an xpub. That's what output descriptors do so well. Descriptors combine complete key derivation information with their associated scripts.

**Reply by Kixunil** [reply]
Note that PSBTs can carry derivation paths, so it may be possible.

**Comment by sebastianmontero** [comment]
Thanks!
So say I have `[0cdb4ee2/48'/0'/0'/2']xpub6EUu26aVpSAyij4Be7Yab3JNNkuFH5a3BH8KRJh4Eo1LBhy6e3i2N8JS1fYAGW9bjRfMCRoWtakd38MuWJCKANRwABPBgSotNjMCa1TFt5a/0/* ` the derivation path would be /48'/0'/0'/2', correct? So with this information would it be possible to derive the public key? Or would I need something else?

**Reply by dpc** [reply]
I'm not an expert here, but AFAIU yes.

**Comment by sebastianmontero** [comment]
Thanks, I was able to derive the public key from the xpub, but additionally I need to specify the child key index.

**Reply by DanGould** [reply]
yes, software will derive public keys by their index. It will continue to derive as long as it finds used keys up to a *gap limit* from the last used key. Typically that limit is 20, but it can vary in software. You will probably have to derive many keys until you find one that matches that psbt if you don't have the original software that created the psbt do it for you

**Comment by sebastianmontero** [comment]
Thank you, yes I was able to find the child key index in the partial signatures of the psbt, the problem I have know is verifying a finalized  psbt, because the partial signatures are not included, does any one know if its possible to get the public keys that signed the psbt from the final script witness? How can I decode the final script witness?

**Reply by Kixunil** [reply]
If you have a specific, known script you should be able to just read specific witness element. The upcoming version of the library will even allow retrieving it directly without iterating (but iterating is often fine anyway).
